docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+  | | Keyhash        | |   <- indicates which public key for sig (struct image_tlv)
+  | +----------------+ |
+  | | Signature      | |   <- signature from A - B(struct image_tlv), only hash
+  | +----------------+ |
+  +--------------------+
+```
+
+But if manifest will be expanded with more info, then manifest itself required
+to be integrity and authenticity protected. The easiest solution described
+below, it was proposed by David Brown. There were alternatives. I chose, which
+has less code impact.
+
+```
+  +--------------------+
+  | Header             |   <- struct image_header
+  +--------------------+
+  | Payload            |
+  +--------------------+
+  | Manifest           |
+  | +----------------+ |
+  | | Manifest header| |   <- struct image_tlv_info
+A | +----------------+ |
+  | | SHA256 hash    | |   <- hash of header and payload (struct image_tlv)
+  | +----------------+ |
+  | | Dependency     | |   <- Dependency entry(struct image_tlv)
+B | +----------------+ |
+  | | Keyhash        | |   <- indicates which public key for sig (struct image_tlv)
+  | +----------------+ |
+  | | Signature      | |   <- signature from A - B(struct image_tlv)
+  | +----------------+ |
+  +--------------------+
+```
+
+# Flash partitioning
+In order to support independent update of multiple images the flash must be
+partitioned further to arrange two slot for each image.
+```
++--------------------+
+| MCUBoot            |
++--------------------+
+      ~~~~~              <- memory might be not contiguous
++--------------------+
+| Image 0            |
+| primary   area     |
++--------------------+
+| Image 0            |
+| secondary area     |
++--------------------+
+      ~~~~~              <- memory might be not contiguous
++--------------------+
+| Image 1            |
+| primary   area     |
++--------------------+
+| Image 1            |
+| secondary area     |
++--------------------+
+| Scratch            |
++--------------------+
+```
+Since this design mostly address the ARMv8-M architecture where the secure and
+non-secure image naming convention is widely used therefore slot can be called
+to:
+  - secure image slot 0
+  - secure image slot 1
+  - non-secure image slot 0
+  - non-secure image slot 1
+
+A more architecture agnostic naming could be this(preferred by me):
+  - Image 0 primary slot
+  - Image 0 secondary slot
+  - Image 1 primary slot
+  - Image 1 secondary slot
+  - ...
+  - Image N primary slot
+  - Image N secondary slot
+
+Mapping of image slots to flash areas might be need to be improved or changed a
+bit, because currently it is a hard coded mapping, with applying an offset.
+
+# Shared scratch
+With the increased partitioning of flash, the overhead of unused flash areas
+caused by padding are also increasing. In order to spare as much flash space as
+possible it could be good approach to share one scratch area among images. The
+swap status info could be saved to image's primary slot or to the common
+scratch.
```

Comment:
`scratch` has to be able to fit the largest sector on the device. This is not a problematic for ortogonal small sector size devices (anything below 4K). But it turns very complicated on many of the STM32s that have 128K sized sectors as well as non-linear distribution of sizes. I tend to think it would be much easier to allow for swap status to remain on the origin slot (both secure and non-secure).

It may be that future devices will tend to go with small sector sizes and linear distribution which would allow us to move swap area to its own partition making this much easier, but I would suggest this being another future task not related to the dual image support.

Reply:
I wouldn't imaging a device with large sectors that also needs multiple images. At least we should document this to try and discourage this kind of configuration.